kvserver: add ff index to TestFlowControlSendQueueRange*

kvoli opened the pull request:
First two commits from #136330
Third commit from #136258

---

kvflowcontrol: add force flush index to inspectz (range) handle

This commit introduces a new field on `kvflowinspectpb.Handle`,
`ForceFlushIndex`.

`ForceFlushIndex` is an index up to (and including) which the range
controller running in pull mode must force-flush all send streams.

This is intended for use in live debugging and asserting on state in
testing.

---

sql: add force_flush_index to internal.kv_flow_control_handles_v2

Add a new column `force_flush_index` to
`crdb_internal.kv_flow_control_handles_v2`.

The column value represents an index up to (and including) which the
range controller running in pull mode must force-flush all send streams.

---

kvserver: add ff index to TestFlowControlSendQueueRangeSplitMerge

Add new echo testdata assertions on the `force_flush_index` of the test
ranges pre-split, post-split and post-merge, in
`TestFlowControlSendQueueRangeSplitMerge`.

---

kvserver: add ff index to TestFlowControlSendQueueRangeMigrate

Add new echo testdata assertions on the force_flush_index of the test
ranges pre-migrate and post-migrate in
`TestFlowControlSendQueueRangeMigrate`.

Part of: #132614
Release note: None

kvoli committed:
kvserver:  add TestFlowControlSendQueueRangeSplitMerge test

Add a new rac2 flow control integration test,
`TestFlowControlSendQueueRangeSplitMerge`.

This test takes the following steps:

```sql
-- We will exhaust the tokens across all streams while admission is blocked on
-- n3, using a single 4 MiB (deduction, the write itself is small) write. Then,
-- we will write a 1 MiB put to the range, split it, write a 1 MiB put to the
-- RHS range, merge the ranges, and write a 1 MiB put to the merged range. We
-- expect that at each stage where a send queue develops n1->s3, the send queue
-- will be flushed by the range merge and range split range operations.
```

Part of: #132614
Release note: None

kvoli committed:
kvserver: add TestFlowControlSendQueueRangeMigrate test

Add a new rac2 flow control integration test,
`TestFlowControlSendQueueRangeMigrate`.

This test takes the following steps:

```sql
-- We will exhaust the tokens across all streams while admission is blocked on
-- n3, using a single 4 MiB (deduction, the write itself is small) write. Then,
-- we will write a 1 MiB put to the range, migrate the range, and write a 1 MiB
-- put to the migrated range. We expect that the migration will trigger a force
-- flush of the send queue.
```

Part of: #132614
Release note: None

sumeerbhola committed:
rac2,replica_rac2: add ForceFlushIndexChangedLocked to Processor, RangeController

This is used to set the highest index up to which all send-queues in
pull mode must be force-flushed.

Informs #135601

Epic: CRDB-37515

Release note: None

sumeerbhola committed:
kvserver: write ForceFlushIndex and integrate with RAC2

- ReplicatedEvalResult.DoTimelyApplicationToAllReplicas is added and set
  in splitTriggerHelper, MergeTrigger, Migrate, Subsume.
- The previous setting to true is gated on cluster version
  V25_1_AddRangeForceFlushKey.
- This causes ReplicaState.ForceFlushIndex to be set, and
  RangeForceFlushKey (a replicated range-id local key) to be written when
  applying the corresponding batch to the state machine. The index is set
  to the index of the entry being applied, and is monotonically
  increasing.
- replica_rac2.Processor and rac2.RangeController have
  existing ForceFlushIndexChangedLocked methods that are called whenever
  the Replica sees a change in the force-flush-index.

Fixes #135601

Epic: CRDB-37515

Release note: None

kvoli committed:
kvflowcontrol: add force flush index to inspectz (range) handle

This commit introduces a new field on `kvflowinspectpb.Handle`,
`ForceFlushIndex`.

`ForceFlushIndex` is an index up to (and including) which the range
controller running in pull mode must force-flush all send streams.

This is intended for use in live debugging and asserting on state in
testing.

Part of: #132614
Release note: None

kvoli committed:
sql: add force_flush_index to internal.kv_flow_control_handles_v2

Add a new column `force_flush_index` to
`crdb_internal.kv_flow_control_handles_v2`.

The column value represents an index up to (and including) which the
range controller running in pull mode must force-flush all send streams.

Part of: #132614
Release note: None

kvoli committed:
kvserver: add ff index to TestFlowControlSendQueueRangeSplitMerge

Add new echo testdata assertions on the `force_flush_index` of the test
ranges pre-split, post-split and post-merge, in
`TestFlowControlSendQueueRangeSplitMerge`.

Part of: #132614
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136512)
<!-- Reviewable:end -->

kvoli committed:
kvserver: add ff index to TestFlowControlSendQueueRangeMigrate

Add new echo testdata assertions on the force_flush_index of the test
ranges pre-migrate and post-migrate in
`TestFlowControlSendQueueRangeMigrate`.

Part of: #132614
Release note: None
